File: backend/app/services/file_service.py
from __future__ import annotations

import asyncio
import hashlib
import io
import re
import threading
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from dataclasses import dataclass
from datetime import UTC, timedelta
from pathlib import Path

from fastapi import UploadFile
from PIL import Image, UnidentifiedImageError
from sqlalchemy import exists, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.sql.elements import ColumnElement

from app.core.config import settings
from app.core.errors import AppError, not_found
from app.core.identifiers import uuid7_string
from app.models import (
    FileObject,
    PurchaseMaterialImage,
    StockMaterialImage,
)
from app.schemas import (
    FileObjectRead,
    OrphanFileCleanupRead,
    OrphanFileRead,
    OrphanFileReportRead,
)
from app.services.common import file_read, utc_aware, utcnow

ACCEPTED_TYPES = {"image/jpeg", "image/png", "image/webp"}
PREVIEW_MIME_TYPE = "image/webp"
MANAGED_FILE_NAME = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\.png$"
)


@dataclass
class DigestLockEntry:
    lock: asyncio.Lock
    users: int = 0


_digest_locks: dict[str, DigestLockEntry] = {}
_digest_locks_guard = threading.Lock()


def file_path(file_id: str) -> Path:
    return settings.upload_dir / f"{file_id}.png"


@asynccontextmanager
async def _digest_lock(digest: str) -> AsyncIterator[None]:
    with _digest_locks_guard:
        entry = _digest_locks.setdefault(digest, DigestLockEntry(lock=asyncio.Lock()))
        entry.users += 1
    try:
        async with entry.lock:
            yield
    finally:
        with _digest_locks_guard:
            entry.users -= 1
            if entry.users == 0:
                _digest_locks.pop(digest, None)


def _unreferenced() -> ColumnElement[bool]:
    return ~exists().where(StockMaterialImage.file_id == FileObject.id) & ~exists().where(
        PurchaseMaterialImage.file_id == FileObject.id
    )


def _managed_disk_files() -> dict[str, Path]:
    settings.upload_dir.mkdir(parents=True, exist_ok=True)
    return {
        path.name: path
        for path in settings.upload_dir.iterdir()
        if path.is_file() and MANAGED_FILE_NAME.fullmatch(path.name)
    }


def _render_preview(source: Path, size: int) -> bytes:
    output = io.BytesIO()
    with Image.open(source) as image:
        image.thumbnail((size, size), Image.Resampling.LANCZOS)
        converted = image.convert("RGBA" if "A" in image.getbands() else "RGB")
        converted.save(output, format="WEBP", quality=82, method=6)
    return output.getvalue()


async def save_image(session: AsyncSession, upload: UploadFile) -> FileObjectRead:
    if upload.content_type not in ACCEPTED_TYPES:
        raise AppError("INVALID_IMAGE_TYPE", "仅支持 JPEG、PNG 或 WebP 图片")
    raw = await upload.read(settings.max_image_bytes + 1)
    if len(raw) > settings.max_image_bytes:
        raise AppError("IMAGE_TOO_LARGE", "单张图片不能超过 10 MB", status_code=413)
    try:
        with Image.open(io.BytesIO(raw)) as source:
            source.load()
            converted = source.convert("RGBA" if "A" in source.getbands() else "RGB")
            width, height = converted.size
            output = io.BytesIO()
            converted.save(output, format="PNG", optimize=True)
    except (UnidentifiedImageError, OSError, Image.DecompressionBombError) as exc:
        raise AppError("INVALID_IMAGE", "图片无法解码") from exc

    data = output.getvalue()
    digest = hashlib.sha256(data).hexdigest()
    async with _digest_lock(digest):
        existing_items = list(
            (
                await session.scalars(
                    select(FileObject)
                    .where(FileObject.sha256 == digest)
                    .order_by(FileObject.created_at, FileObject.id)
                )
            ).all()
        )
        missing_item: FileObject | None = None
        for existing in existing_items:
            existing_path = file_path(existing.id)
            if not existing_path.is_file():
                missing_item = missing_item or existing
                continue
            if existing.size_bytes != len(data):
                continue
            existing_data = await asyncio.to_thread(existing_path.read_bytes)
            if existing_data == data:
                return file_read(existing)

        if missing_item is not None:
            target = file_path(missing_item.id)
            settings.upload_dir.mkdir(parents=True, exist_ok=True)
            await asyncio.to_thread(target.write_bytes, data)
            missing_item.mime_type = "image/png"
            missing_item.size_bytes = len(data)
            missing_item.width = width
            missing_item.height = height
            try:
                await session.commit()
            except Exception:
                await session.rollback()
                await asyncio.to_thread(target.unlink, missing_ok=True)
                raise
            return file_read(missing_item)

        file_id = uuid7_string()
        settings.upload_dir.mkdir(parents=True, exist_ok=True)
        target = file_path(file_id)
        await asyncio.to_thread(target.write_bytes, data)
        item = FileObject(
            id=file_id,
            original_name=Path(upload.filename or "image").name[:255],
            mime_type="image/png",
            size_bytes=len(data),
            width=width,
            height=height,
            sha256=digest,
        )
        session.add(item)
        try:
            await session.flush()
            # 上传是独立事务。只有数据库记录真正提交后才向客户端返回成功。
            await session.commit()
        except Exception:
            await session.rollback()
            await asyncio.to_thread(target.unlink, missing_ok=True)
            raise
        return file_read(item)


async def get_image(session: AsyncSession, file_id: str) -> tuple[FileObject, Path]:
    item = await session.get(FileObject, file_id)
    if item is None:
        raise not_found("图片")
    path = file_path(file_id)
    if not path.is_file():
        raise AppError("FILE_MISSING", "图片文件不存在")
    return item, path


async def render_preview(path: Path, size: int) -> bytes:
    try:
        return await asyncio.to_thread(_render_preview, path, size)
    except (UnidentifiedImageError, OSError, Image.DecompressionBombError) as exc:
        raise AppError("INVALID_IMAGE", "图片无法生成预览") from exc


async def delete_image(session: AsyncSession, file_id: str) -> None:
    item = await session.get(FileObject, file_id)
    if item is None:
        raise not_found("图片")
    referenced = False
    for model in (StockMaterialImage, PurchaseMaterialImage):
        if await session.scalar(select(exists().where(model.file_id == file_id))):
            referenced = True
            break
    if referenced:
        raise AppError("FILE_IN_USE", "图片已被业务引用，不能删除", status_code=409)
    path = file_path(file_id)
    await session.delete(item)
    await session.commit()
    await asyncio.to_thread(path.unlink, missing_ok=True)


async def inspect_orphans(
    session: AsyncSession, older_than_hours: int
) -> OrphanFileReportRead:
    cutoff = utcnow() - timedelta(hours=older_than_hours)
    unreferenced = list(
        (
            await session.scalars(
                select(FileObject)
                .where(FileObject.created_at <= cutoff, _unreferenced())
                .order_by(FileObject.created_at, FileObject.id)
            )
        ).all()
    )
    all_records = list(
        (await session.execute(select(FileObject.id, FileObject.created_at))).all()
    )
    disk_files = await asyncio.to_thread(_managed_disk_files)
    cutoff_timestamp = cutoff.replace(tzinfo=UTC).timestamp()
    record_ids = {file_id for file_id, _ in all_records}
    untracked = sorted(
        name
        for name, path in disk_files.items()
        if name.removesuffix(".png") not in record_ids and path.stat().st_mtime <= cutoff_timestamp
    )
    missing = sorted(
        file_id
        for file_id, created_at in all_records
        if created_at <= cutoff and f"{file_id}.png" not in disk_files
    )
    return OrphanFileReportRead(
        cutoff=utc_aware(cutoff),
        unreferenced_records=[
            OrphanFileRead(
                id=item.id,
                original_name=item.original_name,
                size_bytes=item.size_bytes,
                created_at=utc_aware(item.created_at),
                file_exists=f"{item.id}.png" in disk_files,
            )
            for item in unreferenced
        ],
        untracked_file_names=untracked,
        missing_file_ids=missing,
    )


async def cleanup_orphans(
    session: AsyncSession, older_than_hours: int
) -> OrphanFileCleanupRead:
    report = await inspect_orphans(session, older_than_hours)
    record_ids = [item.id for item in report.unreferenced_records]
    if record_ids:
        records = list(
            (
                await session.scalars(
                    select(FileObject).where(FileObject.id.in_(record_ids), _unreferenced())
                )
            ).all()
        )
        deleted_record_ids = [item.id for item in records]
        for item in records:
            await session.delete(item)
        await session.commit()
    else:
        deleted_record_ids = []

    candidates = [f"{file_id}.png" for file_id in deleted_record_ids]
    candidates.extend(report.untracked_file_names)
    deleted_file_names: list[str] = []
    for name in sorted(set(candidates)):
        path = settings.upload_dir / name
        if path.is_file():
            await asyncio.to_thread(path.unlink)
            deleted_file_names.append(name)
    return OrphanFileCleanupRead(
        cutoff=report.cutoff,
        deleted_record_ids=deleted_record_ids,
        deleted_file_names=deleted_file_names,
    )

File: server/internal/service/file_service.go
// Package service 文件服务：图片上传/读取/预览/孤儿清理（等价 file_service.py）。
package service

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"image"
	"image/color"
	stddraw "image/draw"
	"image/png"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/kolesa-team/go-webp/encoder"
	"github.com/kolesa-team/go-webp/webp"
	xdraw "golang.org/x/image/draw"
	_ "golang.org/x/image/webp"
	"gorm.io/gorm"

	"github.com/yangrucheng/materials-manager/server/internal/config"
	apperrors "github.com/yangrucheng/materials-manager/server/internal/errors"
	"github.com/yangrucheng/materials-manager/server/internal/models"
	"github.com/yangrucheng/materials-manager/server/internal/security"
)

const (
	acceptedJpeg = "image/jpeg"
	acceptedPng  = "image/png"
	acceptedWebp = "image/webp"
	previewMime  = "image/webp"
)

var managedFileName = regexp.MustCompile(
	`^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\.png$`)

var digestLocks sync.Mutex
var digestLockMap = map[string]*digestLockEntry{}

type digestLockEntry struct {
	lock  sync.Mutex
	users int
}

func withDigestLock(digest string, fn func()) {
	digestLocks.Lock()
	entry, ok := digestLockMap[digest]
	if !ok {
		entry = &digestLockEntry{}
		digestLockMap[digest] = entry
	}
	entry.users++
	digestLocks.Unlock()

	entry.lock.Lock()
	defer func() {
		entry.lock.Unlock()
		digestLocks.Lock()
		entry.users--
		if entry.users == 0 {
			delete(digestLockMap, digest)
		}
		digestLocks.Unlock()
	}()
	fn()
}

func filePath(cfg *config.Config, fileID string) string {
	return filepath.Join(cfg.UploadDirPath, fileID+".png")
}

func acceptedImageType(contentType string) bool {
	switch contentType {
	case acceptedJpeg, acceptedPng, acceptedWebp:
		return true
	}
	return false
}

func imageHasAlpha(img image.Image) bool {
	switch img.ColorModel() {
	case color.RGBAModel, color.NRGBAModel, color.AlphaModel,
		color.RGBA64Model, color.NRGBA64Model, color.Alpha16Model:
		return true
	}
	if paletted, ok := img.(*image.Paletted); ok {
		for _, c := range paletted.Palette {
			if _, _, _, a := c.RGBA(); a < 0xffff {
				return true
			}
		}
	}
	return false
}

// decodeAndConvertPNG 解码并转为 PNG（含 alpha 保留 RGBA，否则 RGB）。
func decodeAndConvertPNG(raw []byte) ([]byte, int, int, error) {
	img, _, err := image.Decode(bytes.NewReader(raw))
	if err != nil {
		return nil, 0, 0, err
	}
	bounds := img.Bounds()
	width, height := bounds.Dx(), bounds.Dy()
	hasAlpha := imageHasAlpha(img)
	out := image.NewRGBA(bounds)
	stddraw.Draw(out, bounds, img, bounds.Min, stddraw.Src)
	var target image.Image = out
	if !hasAlpha {
		target = image.NewRGBA(bounds)
		stddraw.Draw(target.(*image.RGBA), bounds, img, bounds.Min, stddraw.Src)
	}
	var buf bytes.Buffer
	enc := png.Encoder{CompressionLevel: png.BestCompression}
	if err := enc.Encode(&buf, target); err != nil {
		return nil, 0, 0, err
	}
	return buf.Bytes(), width, height, nil
}

// SaveImage 保存上传图片：类型校验 -> 转 PNG -> sha256 去重 -> 落盘 + 入库（独立事务）。
func SaveImage(cfg *config.Config, db *gorm.DB, contentType, originalName string, raw []byte) (*models.FileObject, *apperrors.AppError) {
	if !acceptedImageType(contentType) {
		return nil, apperrors.New("INVALID_IMAGE_TYPE", "仅支持 JPEG、PNG 或 WebP 图片", 0, nil)
	}
	if int64(len(raw)) > cfg.MaxImageBytes {
		return nil, apperrors.New("IMAGE_TOO_LARGE", "单张图片不能超过 10 MB", 413, nil)
	}
	data, width, height, err := decodeAndConvertPNG(raw)
	if err != nil {
		return nil, apperrors.New("INVALID_IMAGE", "图片无法解码", 0, nil)
	}
	sum := sha256.Sum256(data)
	digest := hex.EncodeToString(sum[:])
	var result *models.FileObject
	var saveErr error
	withDigestLock(digest, func() {
		result, saveErr = saveImageLocked(cfg, db, data, width, height, digest, originalName)
	})
	if saveErr != nil {
		return nil, DatabaseError(saveErr)
	}
	return result, nil
}

func saveImageLocked(cfg *config.Config, db *gorm.DB, data []byte, width, height int, digest, originalName string) (*models.FileObject, error) {
	var existing []models.FileObject
	if err := db.Where("sha256 = ?", digest).Order("created_at, id").Find(&existing).Error; err != nil {
		return nil, err
	}
	var missingItem *models.FileObject
	for i := range existing {
		item := &existing[i]
		path := filePath(cfg, item.ID)
		if _, statErr := os.Stat(path); os.IsNotExist(statErr) {
			if missingItem == nil {
				missingItem = item
			}
			continue
		}
		if item.SizeBytes != int64(len(data)) {
			continue
		}
		diskData, readErr := os.ReadFile(path)
		if readErr != nil {
			continue
		}
		if bytes.Equal(diskData, data) {
			return item, nil
		}
	}
	_ = os.MkdirAll(cfg.UploadDirPath, 0o755)
	if missingItem != nil {
		target := filePath(cfg, missingItem.ID)
		if err := os.WriteFile(target, data, 0o644); err != nil {
			return nil, err
		}
		updates := map[string]any{
			"mime_type":  "image/png",
			"size_bytes": len(data),
			"width":      width,
			"height":     height,
		}
		if err := db.Model(&models.FileObject{}).Where("id = ?", missingItem.ID).Updates(updates).Error; err != nil {
			_ = os.Remove(target)
			return nil, err
		}
		return missingItem, nil
	}
	fileID := security.UUID7String()
	target := filePath(cfg, fileID)
	if err := os.WriteFile(target, data, 0o644); err != nil {
		return nil, err
	}
	item := models.FileObject{
		ID:           fileID,
		OriginalName: truncateName(originalName),
		MimeType:     "image/png",
		SizeBytes:    int64(len(data)),
		Width:        width,
		Height:       height,
		SHA256:       digest,
		CreatedAt:    models.UTCNow(),
		UpdatedAt:    models.UTCNow(),
		Version:      1,
	}
	if err := db.Create(&item).Error; err != nil {
		_ = os.Remove(target)
		return nil, err
	}
	return &item, nil
}

func truncateName(s string) string {
	base := filepath.Base(s)
	if base == "." || base == "/" || base == "" {
		base = "image"
	}
	if len(base) > 255 {
		base = base[:255]
	}
	return base
}

// GetImage 查询图片记录与磁盘路径。
func GetImage(cfg *config.Config, db *gorm.DB, fileID string) (*models.FileObject, string, *apperrors.AppError) {
	var item models.FileObject
	err := db.Where("id = ?", fileID).First(&item).Error
	if IsNotFound(err) {
		return nil, "", apperrors.NotFound("图片")
	}
	if err != nil {
		return nil, "", DatabaseError(err)
	}
	path := filePath(cfg, fileID)
	if _, statErr := os.Stat(path); os.IsNotExist(statErr) {
		return nil, "", apperrors.New("FILE_MISSING", "图片文件不存在", 0, nil)
	}
	return &item, path, nil
}

// RenderPreview 生成 size 缩略图的 WebP 预览（质量 82，LANCZOS 近似 CatmullRom）。
func RenderPreview(sourcePath string, size int) ([]byte, error) {
	f, err := os.Open(sourcePath)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	src, _, err := image.Decode(f)
	if err != nil {
		return nil, err
	}
	thumb := thumbnail(src, size)
	var buf bytes.Buffer
	hasAlpha := imageHasAlpha(thumb)
	var wimg *image.RGBA
	_ = wimg
	if hasAlpha {
		wimg = thumb.(*image.RGBA)
	} else {
		rgb := image.NewRGBA(thumb.Bounds())
		stddraw.Draw(rgb, thumb.Bounds(), thumb, thumb.Bounds().Min, stddraw.Src)
		wimg = rgb
	}
	opts, err := encoder.NewLossyEncoderOptions(encoder.PresetDefault, 82)
	if err != nil {
		return nil, err
	}
	if err := webp.Encode(&buf, wimg, opts); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// thumbnail 保持比例缩放到不超过 size×size。
func thumbnail(src image.Image, size int) image.Image {
	bounds := src.Bounds()
	w, h := bounds.Dx(), bounds.Dy()
	if w <= size && h <= size {
		return src
	}
	ratio := float64(size) / float64(maxInt(w, h))
	nw := int(float64(w) * ratio)
	nh := int(float64(h) * ratio)
	if nw < 1 {
		nw = 1
	}
	if nh < 1 {
		nh = 1
	}
	dst := image.NewRGBA(image.Rect(0, 0, nw, nh))
	xdraw.CatmullRom.Scale(dst, dst.Bounds(), src, bounds, xdraw.Over, nil)
	return dst
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// IsFileReferenced 判断图片是否被业务引用。
func IsFileReferenced(db *gorm.DB, fileID string) (bool, error) {
	var count int64
	for _, model := range []any{
		&models.StockMaterialImage{}, &models.PurchaseMaterialImage{},
		&models.PurchaseRequestLineImage{},
	} {
		if err := db.Model(model).Where("file_id = ?", fileID).Count(&count).Error; err != nil {
			return false, err
		}
		if count > 0 {
			return true, nil
		}
	}
	return false, nil
}

// DeleteImage 删除图片（被引用则 409 FILE_IN_USE）。
func DeleteImage(cfg *config.Config, db *gorm.DB, fileID string) *apperrors.AppError {
	item, _, appErr := GetImage(cfg, db, fileID)
	if appErr != nil {
		return appErr
	}
	referenced, err := IsFileReferenced(db, fileID)
	if err != nil {
		return DatabaseError(err)
	}
	if referenced {
		return apperrors.New("FILE_IN_USE", "图片已被业务引用，不能删除", 409, nil)
	}
	if err := db.Delete(&models.FileObject{}, "id = ?", item.ID).Error; err != nil {
		return DatabaseError(err)
	}
	_ = os.Remove(filePath(cfg, fileID))
	return nil
}

// OrphanFileRecord 孤儿记录。
type OrphanFileRecord struct {
	ID           string
	OriginalName string
	SizeBytes    int64
	CreatedAt    time.Time
	FileExists   bool
}

// InspectOrphans 孤儿报告：未被引用且早于 cutoff 的记录 + 磁盘上无记录的 .png。
func InspectOrphans(cfg *config.Config, db *gorm.DB, olderThanHours int) (cutoff time.Time, unreferenced []OrphanFileRecord, untracked, missing []string, appErr *apperrors.AppError) {
	cutoff = models.UTCNow().Add(-time.Duration(olderThanHours) * time.Hour)

	var all []models.FileObject
	if err := db.Order("created_at, id").Find(&all).Error; err != nil {
		return time.Time{}, nil, nil, nil, DatabaseError(err)
	}
	unreferenced = []OrphanFileRecord{}
	for i := range all {
		item := &all[i]
		if item.CreatedAt.After(cutoff) {
			continue
		}
		referenced, err := IsFileReferenced(db, item.ID)
		if err != nil {
			return time.Time{}, nil, nil, nil, DatabaseError(err)
		}
		if referenced {
			continue
		}
		path := filePath(cfg, item.ID)
		_, statErr := os.Stat(path)
		unreferenced = append(unreferenced, OrphanFileRecord{
			ID:           item.ID,
			OriginalName: item.OriginalName,
			SizeBytes:    item.SizeBytes,
			CreatedAt:    item.CreatedAt,
			FileExists:   statErr == nil,
		})
	}

	diskFiles := managedDiskFiles(cfg)
	recordIDs := map[string]bool{}
	for _, item := range all {
		recordIDs[item.ID] = true
	}
	cutoffUnix := cutoff.Unix()
	untracked = []string{}
	for name, path := range diskFiles {
		if recordIDs[strings.TrimSuffix(name, ".png")] {
			continue
		}
		if info, err := os.Stat(path); err == nil && info.ModTime().Unix() <= cutoffUnix {
			untracked = append(untracked, name)
		}
	}
	sort.Strings(untracked)

	missing = []string{}
	for _, item := range all {
		if item.CreatedAt.After(cutoff) {
			continue
		}
		if _, ok := diskFiles[item.ID+".png"]; !ok {
			missing = append(missing, item.ID)
		}
	}
	sort.Strings(missing)
	return cutoff, unreferenced, untracked, missing, nil
}

func managedDiskFiles(cfg *config.Config) map[string]string {
	_ = os.MkdirAll(cfg.UploadDirPath, 0o755)
	out := map[string]string{}
	entries, err := os.ReadDir(cfg.UploadDirPath)
	if err != nil {
		return out
	}
	for _, e := range entries {
		if e.IsDir() || !managedFileName.MatchString(e.Name()) {
			continue
		}
		out[e.Name()] = filepath.Join(cfg.UploadDirPath, e.Name())
	}
	return out
}

// CleanupOrphans 清理孤儿记录与磁盘文件。
func CleanupOrphans(cfg *config.Config, db *gorm.DB, olderThanHours int) (cutoff time.Time, deletedRecordIDs, deletedFileNames []string, appErr *apperrors.AppError) {
	cutoff, unreferenced, untracked, _, appErr := InspectOrphans(cfg, db, olderThanHours)
	if appErr != nil {
		return time.Time{}, nil, nil, appErr
	}
	deletedRecordIDs = []string{}
	for _, rec := range unreferenced {
		referenced, err := IsFileReferenced(db, rec.ID)
		if err != nil {
			return time.Time{}, nil, nil, DatabaseError(err)
		}
		if referenced {
			continue
		}
		if err := db.Delete(&models.FileObject{}, "id = ?", rec.ID).Error; err != nil {
			return time.Time{}, nil, nil, DatabaseError(err)
		}
		deletedRecordIDs = append(deletedRecordIDs, rec.ID)
	}
	deletedFileNames = []string{}
	candidates := map[string]bool{}
	for _, id := range deletedRecordIDs {
		candidates[id+".png"] = true
	}
	for _, name := range untracked {
		candidates[name] = true
	}
	names := make([]string, 0, len(candidates))
	for name := range candidates {
		names = append(names, name)
	}
	sort.Strings(names)
	for _, name := range names {
		path := filepath.Join(cfg.UploadDirPath, name)
		if _, err := os.Stat(path); err == nil {
			if err := os.Remove(path); err == nil {
				deletedFileNames = append(deletedFileNames, name)
			}
		}
	}
	return cutoff, deletedRecordIDs, deletedFileNames, nil
}
